hm_fashion_recs/candidates.py

```diff
@@ -0,0 +1,438 @@
+        )
+
+        candidates["strategy"] = "cooc"
+        return candidates.drop_duplicates(ignore_index=True)
+
+    def create_candidates_same_product_code(base_candidates: pd.DataFrame) -> pd.DataFrame:
+        item2item = (
+            items[["item", "product_code"]]
+            .merge(
+                items[["item", "product_code"]].rename({"item": "item_with"}, axis=1),
+                on="product_code",
+            )[["item", "item_with"]]
+            .query("item != item_with")
+            .reset_index(drop=True)
+        )
+
+        candidates = (
+            base_candidates.merge(item2item, on="item")
+            .drop("item", axis=1)
+            .rename(columns={"item_with": "item"})
+        )
+
+        candidates["min_rank_meta"] = candidates.groupby(["user", "item"])["rank_meta"].transform(
+            "min"
+        )
+        candidates = candidates.query("rank_meta == min_rank_meta").reset_index(drop=True)
+
+        base_candidates_columns = [c for c in base_candidates.columns if "_" in c]
+        base_candidates_replace = {c: f"same_product_code_{c}" for c in base_candidates_columns}
+        candidates = candidates.rename(columns=base_candidates_replace)
+
+        candidates["strategy"] = "same_product_code"
+        return candidates.drop_duplicates(ignore_index=True)
+
+    def create_candidates_ohe_distance(
+        target_users: np.ndarray, week_start: int, num_weeks: int, num_items: int
+    ) -> pd.DataFrame:
+        users_with_ohe = users[["user"]].query("user in @target_users")
+        cols = [c for c in items.columns if c.endswith("_idx")]
+        for c in cols:
+            tmp = pd.read_pickle(user_features_path / f"user_ohe_agg_week{week_start}_{c}.pkl")
+            # cs = [c for c in tmp.columns if c.endswith('_mean')]
+            # tmp = tmp[['user'] + cs]
+            users_with_ohe = users_with_ohe.merge(tmp, on="user")
+
+        users_with_ohe = users_with_ohe.dropna().reset_index(drop=True)
+        limited_users = users_with_ohe["user"].values
+
+        recent_items = transactions.query("@week_start <= week < @week_start + @num_weeks")[
+            "item"
+        ].unique()
+        items_with_ohe = pd.get_dummies(items[["item"] + cols], columns=cols)
+        items_with_ohe = items_with_ohe.query("item in @recent_items").reset_index(drop=True)
+        limited_items = items_with_ohe["item"].values
+
+        item_cols = [c for c in items_with_ohe.columns if c != "item"]
+        user_cols = [f"user_ohe_agg_{c}" for c in item_cols]
+        users_with_ohe = users_with_ohe[["user"] + user_cols]
+        items_with_ohe = items_with_ohe[["item"] + item_cols]
+
+        a_users = users_with_ohe.drop("user", axis=1).values.astype(np.float32)
+        a_items = items_with_ohe.drop("item", axis=1).values.astype(np.float32)
+        a_users = np.ascontiguousarray(a_users)
+        a_items = np.ascontiguousarray(a_items)
```

Comment:
since conversion to numpy is inevitable it's better to call `.to_pandas()/.to_numpy()` ourselves so Modin objects won't be trying to pretend they know what is `__array_interface__` and struggling to mimic its behavior.
```suggestion
        a_users = np.ascontiguousarray(a_users.to_pandas())
        a_items = np.ascontiguousarray(a_items.to_pandas())
```

Reply:
In this case user is doing this transformation to feed the data to faiss.

Shouldn't this function just create new array automatically?

Reply:
I'm not sure about this, I think I've seen cases where passing a modin dataframe to a numpy's function/constructor led to that numpy were constantly trying to access array interface fields thinking that it operating with a literal pandas object (that supports an array interface). Modin on its own doesn't implement this array interface natively but defaults to pandas on every `.__array_interface__` call which causes dramatic performance degradation in comparison with `.to_numpy()`.

I've seen this case quite ago, maybe this was already fixed. I'll double-check and get back to it. If this `.__array_interface__` thing is still not fixed I'll raise an issue to make the `np.array(modin_df)` conversion more smoothly.

Reply:
Agree, such cases seem quite typical. Probably https://github.com/modin-project/modin/issues/5423 will be a step in this direction